src/test/kotlin/com/hibob/academy/service/UserServiceTest.kt

```diff
@@ -0,0 +1,154 @@
+        val user = User(
+            id = 1L,
+            name = "Nir",
+            email = "Nir@example.com",
+            password = "password",
+            isEmailVerified = false
+        )
+        whenever(userDao.findById(1)).thenReturn(user)
+        assertThrows(IllegalArgumentException::class.java) { userService.registerUser(user) }
+
+    }
+
+
+    @Test
+    fun `registerUser - user registration fails`() {
+        val user = User(
+            id = 1L,
+            name = "Nir",
+            email = "Nir@example.com",
+            password = "password",
+            isEmailVerified = false
+        )
+        whenever(userDao.findById(2)).thenReturn(null)
+        whenever(userDao.save(user.copy(isEmailVerified = false))).thenReturn(false)
+        assertThrows(IllegalStateException::class.java) { userService.registerUser(user) }
+
+
+    }
+
+    @Test
+    fun `registerUser - Failed to send verification email`() {
+        val user = User(
+            id = 1L,
+            name = "Nir",
+            email = "Nir@example.com",
+            password = "password",
+            isEmailVerified = false
+        )
+        whenever(userDao.findById(2)).thenReturn(user).thenReturn(null)
+        whenever(userDao.save(user.copy(isEmailVerified = false))).thenReturn(true)
+        whenever(emailVerificationService.sendVerificationEmail(user.email)).thenReturn(false)
+        assertThrows(IllegalStateException::class.java) { userService.registerUser(user) }
+    }
+
+    @Test
+    fun `registerUser - Success`() {
+        val user = User(
+            id = 1L,
+            name = "Nir",
+            email = "Nir@example.com",
+            password = "password",
+            isEmailVerified = false
+        )
+        whenever(userDao.findById(2)).thenReturn(user).thenReturn(null)
+        whenever(userDao.save(user.copy(isEmailVerified = false))).thenReturn(true)
+        whenever(emailVerificationService.sendVerificationEmail(user.email)).thenReturn(true)
+        assertTrue(userService.registerUser(user))
+    }
+
+
+    @Test
+    fun `verifyUserEmail should throw exception when user not found`() {
+        val userId = 1L
+        val token = "validToken"
+
+        whenever(userDao.findById(userId)).thenReturn(null)
+
+        assertThrows(IllegalArgumentException::class.java) {
+            userService.verifyUserEmail(userId, token)
+        }
+
+    }
+
+
+    @Test
+    fun `verifyUserEmail should throw exception when email verification fails`() {
+        val user = User(
+            id = 1L,
+            name = "Nir",
+            email = "Nir@example.com",
+            password = "password",
+            isEmailVerified = false
+        )
+        val token = "invalidToken"
+
+        whenever(userDao.findById(user.id)).thenReturn(user)
+        whenever(emailVerificationService.verifyEmail(user.email, token)).thenReturn(false)
+
+        assertThrows(IllegalArgumentException::class.java) {
```

Comment:
assertThrows returns the exception so you can also verify the exception message is correct